test.c

```diff
@@ -59,9 +67,28 @@ void do_accept(struct utcp_connection *nc, uint16_t port) {
 ssize_t do_send(struct utcp *utcp, const void *data, size_t len) {
 	int s = *(int *)utcp->priv;
 	outpktno++;
-	if(outpktno < dropto && outpktno >= dropfrom && drand48() < dropout)
-		return len;
+	if(outpktno >= dropfrom && outpktno < dropto) {
+		if(drand48() < dropout)
+			return len;
+		if(!reorder_data && drand48() < reorder) {
+			reorder_data = malloc(len);
+			reorder_len = len;
+			memcpy(reorder_data, data, len);
+			reorder_countdown = 1 + drand48() * reorder_dist;
+			return len;
+		}
+	}
+
+	if(reorder_data) {
+		if(--reorder_countdown < 0) {
+			total_out += reorder_len;
+			send(s, reorder_data, reorder_len, MSG_DONTWAIT);
+			free(reorder_data);
+			reorder_data = NULL;
```

Comment:
Erm, maybe I'm missing something here, but `reorder_data` is not initialized on line 33, so it may be some undefined pointer. Probably a null pointer when built in debug mode and pretty sure random value when built as a release built. Then on line 73 the if would fail but on line 82 the if would succeed and maybe send out a random amount of random memory.
Fix: initialize `reorder_data` properly.